### ISSUE TITLE
how to change .pcap to the .bag 

### ISSUE THREAD

**Issue by yxxxxx38324** [HesaiTechnology/HesaiLidar-ros#8]
I use the pandar 40 to record the data,but I can't figure out how to get the the .bag format

**Comment by trickyMan** [comment]
rosbag record?

**Comment by TheGreatGalaxy** [comment]
I also have the same problem. I have to record data in windowsOS. Now I record my data in .pcap, but  I have to analyze the record data in ROS in Ubuntu 16.04. How should i to do ? anyone konws ? Thanks!

**Comment by trickyMan** [comment]
@dahewujiang You can play pcap using HesaiLidar-ros , and then , using rosbag record the topic to a rosbag.

**Comment by TheGreatGalaxy** [comment]
I have tried this. But when playing  the record .bag file, it seems not realtime. The playback rate is not constant. How to fix this problem?

**Comment by trickyMan** [comment]
https://github.com/HesaiTechnology/HesaiLidarSDK/blob/d9d5b9a94b74e2c4925311b487cc260d3ceebb3e/include/hesaiLidarSDK_IN.h#L55

try to decrease the interval time of reading packets to increase the framerate.

**Comment by TheGreatGalaxy** [comment]
Sorry. I am not understand you well. What do you mean that "try to decrease the interval time of reading packets to increase the framerate."? Which parameters should I modify? Do you think that the unconscious playback rate is caused  by playback ? But I think it caused by record. Because I play the .bag file which recording directly in ROS is right.

**Comment by trickyMan** [comment]
https://github.com/HesaiTechnology/HesaiLidarSDK/blob/d9d5b9a94b74e2c4925311b487cc260d3ceebb3e/include/hesaiLidarSDK_IN.h#L55
```
// 10000 us  = 100 ms = 0.1s per packet
#define HesaiLidarSDK_PCAP_TIME_INTERVAL (100000 / 3000)
```
The pcap reading task will delay a time between 2 packets, You can set this interval in source code. By default , there are 3000 packets in 100ms in Signle Return mode for Pandar40 , so the interval will be 100000us / 3000.

If the return mode is Dual Return , there will be 6000 packets.

**Comment by ADoranth** [comment]
Hello，I did it like what you said, but it still didn't work, the bag data always faster than pcap data.

**Comment by ADoranth** [comment]
@trickyMan we tried many times with different values （100000/3000）, sometimes the  bag played very fast, sometimes very slow, the speed of played bag file didn't steady. I thought that it doesn't steady when we replay the pcap file ,so it doesn't steady when we record it. Can you help us to solve the problem as soon as possible,please. Thanks very much~

**Comment by trickyMan** [comment]
try (100000/180)？ 180 means 180 packets per cycle. or  (100000/300).

**Comment by ADoranth** [comment]
@trickyMan Thanks a lot, I tried, these two pair value（100000/180）and（100000/300）seems much slower than pcap data,  so I changed to （100000/420），it seems much better，does these values all right？

**Comment by trickyMan** [comment]
That’s ok.

**Comment by ADoranth** [comment]
All right,thanks, but now we can't promise that the bag data and the pcap data can absolutely  synchronize, how can we make the beginning and the end of both pcap and the bag data are the same, is there are some codes I can write to come true these?

**Comment by trickyMan** [comment]
Oh, yeah, Every packet in pcap file has it's own timestamp, You can control the delay time using the exact timestamp gap between 2 packets.

**Comment by ADoranth** [comment]
Oh, I got that ,  but i am thinking about how should  I get these timestamp gap?

**Comment by trickyMan** [comment]
refer to the docs of libpcap?

**Comment by ADoranth** [comment]
@trickyMan Yeah, you said that every packet in pcap file has it's own timestamp, I can control the delay time using the exact timestamp gap between 2 packets, so I want to know that is there any codes I can use directly to get the exact timestamp gap between 2 packets and control the delay time?

**Comment by trickyMan** [comment]
@ADoranth 

[Here is the sample code](https://github.com/HesaiTechnology/pandar-ros/blob/master/pandar_grabber/src/pandar_grabber.cpp#L665)

The timestamps are stored in the header of the pcap packet.

**Comment by ADoranth** [comment]
@trickyMan 
All right ,thank you very much , I will try ,but I think it will  costs much time~
